includes/class-wc-payments-express-checkout-button-display-handler.php

```diff
@@ -116,18 +116,32 @@ public function is_woopay_enabled() {
 	 * Add the Pay for order params to the JS config.
 	 */
 	public function add_pay_for_order_params_to_js_config() {
+		global $wp;
+		$order_id = $wp->query_vars['order-pay'];
+		$order    = wc_get_order( $order_id );
+
 		// phpcs:disable WordPress.Security.NonceVerification.Recommended
-		if ( isset( $_GET['pay_for_order'] ) && isset( $_GET['key'] ) ) {
-			global $wp;
-			$order_id = $wp->query_vars['order-pay'];
-			$order    = wc_get_order( $order_id );
+		if ( isset( $_GET['pay_for_order'] ) && isset( $_GET['key'] ) && current_user_can( 'pay_for_order', $order_id ) ) {
 			add_filter(
 				'wcpay_payment_fields_js_config',
 				function( $js_config ) use ( $order ) {
+					$session       = wc()->session;
+					$session_email = '';
+
+					if ( is_a( $session, WC_Session::class ) ) {
+						$customer      = $session->get( 'customer' );
+						$session_email = is_array( $customer ) && isset( $customer['email'] ) ? $customer['email'] : '';
+					}
+
+					$user_email = sanitize_email( wp_unslash( filter_input( INPUT_POST, 'email' ) ) ) ?? $session_email;
+
```

Comment:
Do we still need to set `session_email` after the latest changes? or can we just set it to an empty string if the user does not have the correct access rights?

Reply:
We are not checking the access rights here.  This function only adds the email to the `user_email`.  Let me know if I understood you correctly.

Reply:
Hi Emma, what I don't quite understand is at occasions do we fallback to `user_email`. Could you give me a test where `user_email` is used instead of the `get_billing_email` in lines 146-149?
I've been testing with guest and logged in orders but have not been successful in finding an occasion where `user_email` is required. I am likely missing something and will appreciate your help in testing it out.

Reply:
This is used whenever we have guest orders.  To reproduce:

1. Create a guest order
2. Use incognito or log out of WP
3. Add `error_log($user_email);` right after `$user_email`
4. Visit the pay for order page.  You should see the email input form
5. Enter your email
6. See the debug log with the user email

<img width="1110" alt="Screen Shot 2023-10-30 at 2 42 02 PM" src="https://github.com/Automattic/woocommerce-payments/assets/56378160/c7a40b74-0f34-45b3-be35-c7f0998da7f3">

Reply:
Thanks. But I am more confused now 😅 
The verification in your screenshot is done by WC-core. It will always check against the order's billing email. 

`js_config['billing_email']` we are setting in this PR is only for WooPay's consumption.

Once a user gets pass the verification in the screenshot above, we can be reasonably confident that it's the correct shopper.  So shouldn't we always be setting `$js_config['billing_email']  = $order->get_billing_email()`?

Is there a particular reason why WooPay require `user_email` as the `billing_email` for logged out users?

---
On a separate note, following your test instructions, `$user_email` is an empty string for me when logged out. This is because `sanitize_email( wp_unslash( filter_input( INPUT_POST, 'email' ) ) )` is an empty string and the null coalescent operator is happy to use it without falling back to the value of `$session_email`

Reply:
>The verification in your screenshot is done by WC-core. It will always check against the order's billing email.

This is true.  But we don't have any way of knowing if they are past this check or not.

>Once a user gets pass the verification in the screenshot above, we can be reasonably confident that it's the correct shopper. So shouldn't we always be setting $js_config['billing_email']  = $order->get_billing_email()?

We are not sure if we are through this check or not, they are both on the same page and the core function for this check is not exposed.  You can see that `wcpayConfig ` is visible in your console even before a logged-out user enters their email.

The core check also just uses `sanitize_email( wp_unslash( filter_input( INPUT_POST, 'email' ) ) )`.

>Is there a particular reason why WooPay require user_email as the billing_email for logged-out users?

I'm not quite sure I understand what you mean.  We are using the same POST input from the form that core uses to check for logged-out users.

>On a separate note, following your test instructions, $user_email is an empty string for me when logged out.

Is this after you entered your email in the above form and clicked "Verify"?  I'm able to see the email I entered.

Reply:
@malithsen Is this line why you think we are past the user email check?  

https://github.com/Automattic/woocommerce-payments/blob/0593b5ee805c5eb0906b9651edb559ce163bd37f/includes/class-wc-payments-express-checkout-button-display-handler.php#L129

If so, this is not related to the core email verification.  `current_user_can( 'pay_for_order', $order_id )` always returns `true` for guest orders even when a user is logged out and not verified.

Reply:
Thanks for walking me through this @hsingyuc this makes sense to me now. I misunderstood how the verification works. This looks good :shipit:

Reply:
Thank you for double-checking on this, @malithsen! 👍